test/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule.Tests/ActivityHelperTest.cs

```diff
@@ -70,19 +71,70 @@ public void Has_Started_Returns_Correctly()
         }
 
         [Fact]
-        public void Can_Restore_Activity()
+        public async Task Can_Restore_Activity()
         {
             this.EnableListener();
             var context = HttpContextHelper.GetFakeHttpContext();
-            var rootActivity = ActivityHelper.StartAspNetActivity(this.noopTextMapPropagator, context, null);
+            using var rootActivity = ActivityHelper.StartAspNetActivity(this.noopTextMapPropagator, context, null);
             rootActivity.AddTag("k1", "v1");
             rootActivity.AddTag("k2", "v2");
 
-            Activity.Current = null;
+            Task testTask;
+            using (ExecutionContext.SuppressFlow())
+            {
+                testTask = Task.Run(() =>
+                {
+                    Task.Yield();
+
+                    Assert.Null(Activity.Current);
+
+                    ActivityHelper.RestoreContextIfNeeded(context);
+
+                    Assert.Same(Activity.Current, rootActivity);
+                });
+            }
+
+            await testTask.ConfigureAwait(false);
+        }
+
+        [Fact]
+        public async Task Can_Restore_Baggage()
+        {
+            this.EnableListener();
+
+            var requestHeaders = new Dictionary<string, string>
+            {
+                { BaggageHeaderName, BaggageInHeader },
+            };
+
+            var context = HttpContextHelper.GetFakeHttpContext(headers: requestHeaders);
+            using var rootActivity = ActivityHelper.StartAspNetActivity(new CompositeTextMapPropagator(new TextMapPropagator[] { new TraceContextPropagator(), new BaggagePropagator() }), context, null);
+
+            rootActivity.AddTag("k1", "v1");
+            rootActivity.AddTag("k2", "v2");
+
+            Task testTask;
+            using (ExecutionContext.SuppressFlow())
+            {
+                testTask = Task.Run(() =>
+                {
+                    Task.Yield();
+
+                    Assert.Null(Activity.Current);
```

Comment:
@stephentoub `Activity.Current` uses an `AsyncLocal`. Occasionally this test fails only on net461. Can you Miyagi me on this and point me in the right direction? There is a similar test right above which I haven't seen fail.

What I'm trying to do is get some code running on a thread where ExecutionContext has no local values so I can verify my code properly restores things when we run into that situation. We see this in the wild when running on IIS and it decides to jump threads.

Reply:
@grabyourpitchforks Hey anything jump out at you on this test? Sorry to bug, I'm scratching my head on this one.